### PR TITLE
Add OIDC connection-delay property

### LINKED ISSUE #4619: Quarkus OIDC shouldn't fail, when Auth Server is not up
**Describe the bug**
When I start a quarkus application with the oidc extension used, the startup will fail, when the auth server is not reachable. 

**Expected behavior**
Application should start normal, maybe with a warning and an automatic retry

**Actual behavior**
Application stops with following exception:
`java.util.concurrent.CompletionException: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8180
	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
	at io.quarkus.oidc.VertxKeycloakRecorder.setup(VertxKeycloakRecorder.java:59)
	at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup28.deploy_0(VertxKeycloakBuildStep$setup28.zig:92)
	at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup28.deploy(VertxKeycloakBuildStep$setup28.zig:36)
	at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:137)
	at io.quarkus.runtime.Application.start(Application.java:94)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:135)
	at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:180)
	at io.quarkus.dev.DevModeMain.start(DevModeMain.java:94)
	at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8180
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:336)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:685)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:549)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:511)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
2019-10-16 20:28:40,692 ERROR [io.qua.dev.DevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:185)
	at io.quarkus.runtime.Application.start(Application.java:94)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:135)
	at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:180)
	at io.quarkus.dev.DevModeMain.start(DevModeMain.java:94)
	at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: java.util.concurrent.CompletionException: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8180
	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:412)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2044)
	at io.quarkus.oidc.VertxKeycloakRecorder.setup(VertxKeycloakRecorder.java:59)
	at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup28.deploy_0(VertxKeycloakBuildStep$setup28.zig:92)
	at io.quarkus.deployment.steps.VertxKeycloakBuildStep$setup28.deploy(VertxKeycloakBuildStep$setup28.zig:36)
	at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:137)
	... 5 more
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:8180
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779)
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:327)
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:336)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:685)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:632)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:549)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:511)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)`

**To Reproduce**
1. Build minimal application
2. Add oicd dependency
3. Add auth server address to application.properties
4. Start server

**Configuration**
```properties
# Add your application.properties here, if applicable.

`quarkus.oidc.auth-server-url=http://localhost:8180/auth/realms/test
quarkus.oidc.client-id=test-client`

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #4619

### CODE REVIEW

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfig.java [hunk 1]
```diff
@@ -50,6 +51,13 @@
     @ConfigItem
     Optional<String> clientId;
 
+    /**
+     * The time the adapter can wait for before connecting to the OIDC server.
```
Comment (gsmet) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfig.java]:
```suggestion
     * The time the adapter can wait for before connecting to the OIDC server.
```
For now it seems to be the time the adapter **will** wait. Not sure if we want that or if we want to retry from time to time?

Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfig.java]: @gsmet I've updated the code to retry every 2 seconds for up to the` connectionDelay` time in seconds, by default or if this property is set to less than 2 secs it is just a single connection attempt

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java [hunk 2]
```diff
@@ -45,19 +49,43 @@ public void setup(OidcConfig config, RuntimeValue<Vertx> vertx, BeanContainer be
...
+                        }
+                    }
+                });
+
+                auth = cf.join();
+                break;
+            } catch (OIDCException ex) {
+                if (i + 1 < connectionRetryCount) {
+                    try {
+                        Thread.sleep(2000);
```
Comment (stuartwdouglas) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java]: I wonder if we actually want a more generic approach to dealing with this, as there are quite a few different considerations here:
- Should we use a similar approach for all resources (database connections etc). Would some kind of unified 'resource timeout' parameter make more sense
- How does this affect liveliness/readiness probes. AFAIK if we are trying to connect to the OIDC server then we should be live but not ready
- What happens if the OIDC server goes down while we are running?

This is probably actually a good topic for the F2F.

**Review by stuartwdouglas — APPROVED** [review]
This looks good, but I think we will probably want a more generic solution at some point.

**Comment by UnvirtualHH** [comment]
I tried it in 1.1.0.CR1 but it seems not working. It never goes into the retry, even if it logs that it will, at the start of the application. Maybe u can try it yourself again and give me a feedback if this feature works as expected.

**Pull request merged**